File: SortTest/src/com/renho/Quick.java
package com.renho;

/**
 * 快速排序
 * 递归实现 
 * @author renho
 *
 */
public class Quick {
	
	public int[] sort(int[] array) {
		quickSort(array, 0, array.length - 1);
		return array;
	}
	
	public void quickSort(int[] array, int low, int hight) {
		if(low < hight) {
			int i = low;
			int j = hight;
			int value = array[low];
			while(i < j) {
				while (i < j && array[j] >= value) {
					j--;
				}
				array[i] = array[j];
				while (i < j && array[i] <= value) {
					i++;
				}
				array[j] = array[i];
			}
			array[i] = value;
			quickSort(array, low, i);
			quickSort(array, i + 1, hight);
		}
	}
}

File: ideaworkspace/designpatterns/src/main/scala/com/renho/sort/Quick.scala
package com.renho.sort

/**
  * Created by renho on 2016/7/12.
  */
object Quick {
	def sort(start: Int, end: Int, arr: Array[Int]): Unit ={
		if(start < end) {
			val temp = arr(start)
			var i = start
			var j = end
			while(i < j) {
				while(i<j && arr(j) >= temp) {
					j -= 1
				}
				arr(i) = arr(j)
				while(i<j && arr(i) <= temp) {
					i += 1
				}
				arr(j) = arr(i)
			}
			arr(i) = temp
			sort(start, i, arr)
			sort(i+1, end, arr)
		}
	}
}